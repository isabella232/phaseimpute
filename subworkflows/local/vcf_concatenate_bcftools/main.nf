include { BCFTOOLS_CONCAT } from '../../../modules/nf-core/bcftools/concat'

workflow VCF_CONCATENATE_BCFTOOLS {

    take:
    ch_vcf_tbi // channel: [ [id, panel, chr, tools], vcf, tbi ]

    main:

    ch_versions = Channel.empty()

    // Keep only id from meta
    ch_vcf_tbi_grouped = ch_vcf_tbi
        .map{ metaIPTC, vcf, tbi -> [metaIPTC.subMap("id", "tools", "panel", "batch") + ["chr": "all"], vcf, tbi] }
        .groupTuple( by:0 )
        .map{ metaIPTC, vcf, tbi -> [metaIPTC, vcf, tbi, vcf.size() ] } // Compute number of records
        .branch{
            one: it[3] == 1
            more: it[3] > 1
        }

    // Ligate and concatenate chunks
    BCFTOOLS_CONCAT(ch_vcf_tbi_grouped.more.map{ [it[0], it[1], it[2]] })
    ch_versions = ch_versions.mix(BCFTOOLS_CONCAT.out.versions.first())

    // Join VCFs and TBIs
    ch_vcf_tbi_concat = BCFTOOLS_CONCAT.out.vcf
        .join(BCFTOOLS_CONCAT.out.tbi)

    ch_vcf_tbi_join = ch_vcf_tbi_grouped.one
        .map{ [it[0], it[1][0], it[2][0]] }
        .mix(ch_vcf_tbi_concat)

    emit:
    vcf_tbi      = ch_vcf_tbi_join // channel: [ [id], vcf, tbi ]
    versions     = ch_versions     // channel: [ versions.yml ]
}
