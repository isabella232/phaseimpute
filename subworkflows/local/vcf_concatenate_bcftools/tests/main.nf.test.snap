{
    "Concat content": {
        "content": [
            [
                "VcfFile [chromosomes=[chr21, chr22], sampleCount=3196, variantCount=1739, phased=true, phasedAutodetect=true]"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-18T11:56:39.376973769"
    },
    "Concatenate vcf": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "all"
                        },
                        "1000GP_all.vcf.gz:md5,fc8f8ceff29246d96337302072f8a3bb",
                        "1000GP_all.vcf.gz.tbi:md5,e7180bb953d2bd657c420a5f76a7164d"
                    ]
                ],
                "1": [
                    "versions.yml:md5,0370775d4b565f1cf95dcbea437fa36a"
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "all"
                        },
                        "1000GP_all.vcf.gz:md5,fc8f8ceff29246d96337302072f8a3bb",
                        "1000GP_all.vcf.gz.tbi:md5,e7180bb953d2bd657c420a5f76a7164d"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0370775d4b565f1cf95dcbea437fa36a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-06T15:34:39.215613472"
    }
}